### PR TITLE
Promote egress gateway to stable

### PR DESCRIPTION
**Pull request by jibi**
See individual commits.

### CODE REVIEW

**Comment by jibi** [comment]
Opening this up for review, depending on how long it takes I will push the docs updates (still WIP) either here or in a new PR

File: pkg/egressgateway/endpoint.go [hunk 1]
```diff
@@ -7,10 +7,10 @@ import (
 	"fmt"
 	"net"
 
-	"k8s.io/apimachinery/pkg/types"
-
 	k8sTypes "github.com/cilium/cilium/pkg/k8s/types"
 	"github.com/cilium/cilium/pkg/labels"
+
+	"k8s.io/apimachinery/pkg/types"
```
Comment (tklauser) [inline pkg/egressgateway/endpoint.go]: Nit: I think the order of the import groups should stay as is, i.e. 1) standard library, 2) third-party and 3) local packages. See #18357 for context.

File: pkg/egressgateway/manager.go [hunk 2]
```diff
@@ -4,16 +4,19 @@
...
-
 	k8sTypes "github.com/cilium/cilium/pkg/k8s/types"
 	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/logging"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/maps/egressmap"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
```
Comment (tklauser) [inline pkg/egressgateway/manager.go]: Same comment about import group ordering applies here.

File: pkg/egressgateway/manager.go [hunk 3]
```diff
@@ -4,16 +4,21 @@
...
 	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/logging"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/maps/egressmap"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/option"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+	"k8s.io/apimachinery/pkg/types"
```
Comment (tklauser) [inline pkg/egressgateway/manager.go]: Same comment about import group order applies here.

File: pkg/egressgateway/policy.go [hunk 4]
```diff
@@ -7,18 +7,51 @@ import (
...
 	v2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
 	k8sLabels "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/labels"
 	slim_metav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
 	"github.com/cilium/cilium/pkg/logging/logfields"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
 	"github.com/cilium/cilium/pkg/policy"
 	"github.com/cilium/cilium/pkg/policy/api"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
```
Comment (tklauser) [inline pkg/egressgateway/policy.go]: Same comment about import group order applies here.

**Review by tklauser — COMMENTED** [review]
Some nit about import group ordering inline. Not sure why golangci-lint is not complaining about these given https://github.com/cilium/cilium/blob/f1b52db3374f18d6a0da695198bd8388fe4b3dae/.golangci.yaml#L54-L55

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml [hunk 5]
```diff
@@ -166,6 +166,83 @@ spec:
...
+                    type: string
+                  interface:
+                    description: "Interface is the network interface to which the
+                      egress IP is assigned. \n When none of the Interface or EgressIP
+                      fields is specified, the policy will use the first IPv4 assigned
+                      to the interface with the default route."
+                    type: string
+                  nodeSelector:
```
Comment (julianwiedmann) [inline pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:202]: Should the `nodeSelector` be marked as a required property? Or is a config with only `egressIP` valid as well? 
(effectively that would be the same as using `egressSourceIP` ...?)

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml [hunk 6]
```diff
@@ -176,7 +253,6 @@ spec:
             required:
             - destinationCIDRs
             - egress
-            - egressSourceIP
```
Comment (julianwiedmann) [inline pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:179]: Should we prevent users from building a policy with no `egressSourceIP` and no `egressGateway`? Using `oneOf` maybe?

File: Documentation/cmdref/cilium_bpf_egress.md [hunk 7]
```diff
@@ -21,8 +21,5 @@ Manage the egress routing rules
 ### SEE ALSO
 
 * [cilium bpf](../cilium_bpf)	 - Direct access to local BPF maps
-* [cilium bpf egress delete](../cilium_bpf_egress_delete)	 - Delete egress entries
-* [cilium bpf egress get](../cilium_bpf_egress_get)	 - Get egress entries
 * [cilium bpf egress list](../cilium_bpf_egress_list)	 - List egress policy entries
-* [cilium bpf egress update](../cilium_bpf_egress_update)	 - Update egress entries
```
Comment (julianwiedmann) [inline Documentation/cmdref/cilium_bpf_egress.md:25]: > but now it doesn't really make sense to configure a policy manually through the CLI

because the policy is getting too complex for a CLI-based configuration?

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml (hunk 5 above)
Comment (jibi) [reply pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:202]: It should be marked as required :+1: (I thought it was already by default but apparently it's not)

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml (hunk 6 above)
Comment (jibi) [reply pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:179]: I'm already enforcing that in the logic that parses the policy:
https://github.com/cilium/cilium/blob/c2e37d36f3ba58665fb735d284e93c03c7bf275b/pkg/egressgateway/policy.go#L227-L232 

~but I should probably do it using the `oneOf` annotation in the type definition :+1:~
looks like there's been some discussion around that (https://github.com/kubernetes-sigs/controller-tools/issues/461) but it's not been implemented yet :disappointed:

File: Documentation/cmdref/cilium_bpf_egress.md (hunk 7 above)
Comment (jibi) [reply Documentation/cmdref/cilium_bpf_egress.md:25]: > because the policy is getting too complex for a CLI-based configuration?

not really, in the end every policy can be distilled down to a list of `(source IP, dest CIDR, gateway IP, egress IP)` entries in the egress policy bpf map (and those commands were used exactly for this, to interact just with the bpf map).

The problem is that after the initial implementation we changed the way the manager reacts to events. Now it:
* keeps an internal state of all the objects it is interested in (nodes, endpoints, egress policies)
* whenever there's a new event (e.g. a new endpoint is added or a policy is updated) it updates its internal state to reflect that
* finally, it [reconciles](https://github.com/cilium/cilium/blob/4b1cb31e0c32e956ebd673a847a02cba63700779/pkg/egressgateway/manager.go#L371-L391) the state of the system with its internal state (so it distills the list of nodes, ep and policies down to a list of entries for the bpf map, adds the ones that are missing and removes the ones that should not be present anymore)

Because of this we should not allow a user to manually interact with the map entries, since whenever there's a new event (for example a new endpoint added) the manager will run its reconciliation logic, and will just delete (or add back) the entries that the user manually created (or deleted)

File: pkg/k8s/watchers/cilium_node.go [hunk 8]
```diff
@@ -74,7 +78,10 @@ func (k *K8sWatcher) ciliumNodeInit(ciliumNPClient *k8s.K8sCiliumClient, asyncCo
 							}
 							n := nodeTypes.ParseCiliumNode(ciliumNode)
 							errs := k.CiliumNodeChain.OnUpdateCiliumNode(oldCN, ciliumNode, swgNodes)
-							if isLocal {
```
Comment (julianwiedmann) [inline pkg/k8s/watchers/cilium_node.go:77]: This change from `isLocal` to `n.IsLocal()` feels unrelated ... ?

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml (hunk 6 above)
Comment (julianwiedmann) [reply pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:179]: Looks like we have support for `oneOf` in [our fork](https://github.com/cilium/controller-tools) though :grin:  (commit ad007ee307a33bc6d45aafd383463c6dafd6e746).

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml [hunk 9]
```diff
@@ -40,6 +40,15 @@ spec:
...
           spec:
+            oneOf:
+            - properties:
+                egressSourceIP: {}
+              required:
+              - egressSourceIP
+            - properties:
+                egressGateway: {}
+              required:
+              - egressGateway
```
Comment (pchaigno) [inline pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:51]: Could you add an example in `examples/kubernetes-egress-gateway/`? Mostly interested in that to understand how the structure is changing, but probably also useful to have for documentation purposes.

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml [hunk 10]
```diff
@@ -166,6 +175,85 @@ spec:
...
+                  traffic.
+                properties:
+                  egressIP:
+                    description: "EgressIP is a source IP address that the egress
+                      traffic is redirected to and SNATed with. \n Example: When it
+                      is set to \"192.168.1.100\", matched egress packets will be
+                      redirected to node with IP 192.168.1.100 and SNAT’ed with IP
+                      address 192.168.1.100. \n When none of the Interface or EgressIP
+                      fields is specified, the policy will use the first IPv4 assigned
+                      to the interface with the default route."
```
Comment (pchaigno) [inline pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml]: If I understand correctly, `egressGateway.egressIP` is equivalent to `egressSourceIP`? Are we planning to deprecate `egressSourceIP` in a future version of the CRD?

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml [hunk 11]
```diff
@@ -166,6 +175,85 @@ spec:
                       type: object
                   type: object
                 type: array
+              egressGateway:
+                description: EgressGateway is the gateway node responsible for SNATing
+                  traffic.
+                properties:
+                  egressIP:
+                    description: "EgressIP is a source IP address that the egress
+                      traffic is redirected to and SNATed with. \n Example: When it
```
Comment (pchaigno) [inline pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml]: What happens if the `egressIP` and the `nodeSelector` (and/or the `interface`) don't point to the same IP or even to the same node?

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml [hunk 12]
```diff
@@ -166,6 +175,85 @@ spec:
...
+                    description: "Interface is the network interface to which the
+                      egress IP is assigned. \n When none of the Interface or EgressIP
+                      fields is specified, the policy will use the first IPv4 assigned
+                      to the interface with the default route."
+                    type: string
+                  nodeSelector:
+                    description: This is a label selector which selects the node that
+                      should act as egress gateway for the given policy. In case multiple
+                      nodes are selected, only the first one in the lexicograph ordering
+                      will be used. This field follows standard label selector semantics.
```
Comment (pchaigno) [inline pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml]: > lexicograph ordering

lexical ordering?

Worth stating that it's over node names?

File: Documentation/gettingstarted/egress-gateway.rst [hunk 13]
```diff
@@ -7,11 +7,9 @@
 .. _egress-gateway:
 
 **********************
-Egress Gateway (beta)
+Egress Gateway
 **********************
 
-.. include:: ../beta.rst
-
```
Comment (pchaigno) [inline Documentation/gettingstarted/egress-gateway.rst]: Not sure we want to backport this change to v1.11. Are we about to say the egress gateway is now stable in v1.11 as well? Then, what are we announcing in the v1.12 release blog post? We usually (e.g., host firewall, IPsec) mark the feature stable only in the documentation for the release where it became stable.

Comment (pchaigno) [reply Documentation/gettingstarted/egress-gateway.rst]: Also, shouldn't this change be in its own commit as the very least?

File: pkg/egressgateway/manager.go [hunk 14]
```diff
@@ -159,65 +168,75 @@ func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
...
+	defer manager.mutex.Unlock()
+	delete(manager.nodeDataStore, node.Name)
+	manager.onChangeNodeLocked()
+}
+
+func (manager *Manager) onChangeNodeLocked() {
+	manager.nodes = []nodeTypes.Node{}
+	for _, n := range manager.nodeDataStore {
+		manager.nodes = append(manager.nodes, n)
+	}
```
Comment (pchaigno) [inline pkg/egressgateway/manager.go:225]: I don't understand the need for `manager.nodeDataStore`. Couldn't it be replaced by `manager.nodes` directly?

File: pkg/egressgateway/manager.go [hunk 15]
```diff
@@ -159,65 +168,75 @@ func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
...
+		if err := egressmap.EgressPolicyMap.Update(endpointIP, *dstCIDR, gwc.egressIP.IP, gwc.gatewayIP); err != nil {
+			logger.WithError(err).Error("Error applying egress gateway policy")
+		} else {
+			logger.Info("Egress gateway policy applied")
 		}
 	}
+
+	for _, policyConfig := range manager.policyConfigs {
+		policyConfig.forEachEndpointAndDestination(manager.epDataStore, addEgressRule)
+	}
```
Comment (pchaigno) [inline pkg/egressgateway/manager.go:359]: I would have been nice to move this bit of refactoring to its own commit.

File: pkg/egressgateway/manager.go [hunk 16]
```diff
@@ -227,26 +246,21 @@ func (manager *Manager) removeUnusedEgressRules() {
...
-				}
-
-				for _, endpointIP := range endpoint.ips {
-					for _, dstCIDR := range policyConfig.dstCIDRs {
-						if policyKey.Match(endpointIP, dstCIDR) &&
-							policyVal.Match(policyConfig.egressIP, policyConfig.egressIP) {
-							continue nextPolicyKey
-						}
-					}
-				}
```
Comment (pchaigno) [inline pkg/egressgateway/manager.go:276]: Ditto here regarding refactoring changes.

File: pkg/egressgateway/net.go [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+		return net.IPNet{}, 0, err
+	}
+
+	for _, addr := range addrs {
+		if addr.IP.To4() != nil {
+			return *addr.IPNet, dev.Attrs().Index, nil
+		}
+	}
+
+	return net.IPNet{}, 0, fmt.Errorf("no IPv4 address assigned to interface")
```
Comment (pchaigno) [inline pkg/egressgateway/net.go:36]: Nit:
```suggestion
	return net.IPNet{}, 0, fmt.Errorf("no IPv4 address assigned to interfaces")
```

File: pkg/egressgateway/manager.go [hunk 18]
```diff
@@ -281,6 +373,17 @@ func (manager *Manager) reconcile() {
 		return
 	}
 
+	manager.regenerateGatewayConfigs()
+
+	if option.Config.InstallEgressGatewayRoutes {
+		shouldRetry := manager.addMissingIpRulesAndRoutes()
+		manager.removeUnusedIpRulesAndRoutes()
+
+		if shouldRetry {
```
Comment (nebril) [inline pkg/egressgateway/manager.go:414]: Is one retry always enough?

**Review by nebril — APPROVED** [review]
Looks good overall, one question left inline, didn't review BPF bits.

File: pkg/egressgateway/manager.go (hunk 15 above)
Comment (pchaigno) [inline pkg/egressgateway/manager.go:359]: This is also only refactor AFAICS, right?

File: pkg/egressgateway/policy.go [hunk 19]
```diff
@@ -62,6 +222,29 @@ func ParsePolicy(cenp *v2alpha1.CiliumEgressNATPolicy) (*PolicyConfig, error) {
...
 	egressIP := net.ParseIP(cenp.Spec.EgressSourceIP).To4()
 
+	egressGateway := cenp.Spec.EgressGateway
+
+	switch {
+	case egressIP != nil && egressGateway != nil:
+		return nil, fmt.Errorf("CiliumEgressNATPolicy cannot have both EgressSourceIP and EgressGateway set")
+	case egressIP == nil && egressGateway == nil:
+		return nil, fmt.Errorf("CiliumEgressNATPolicy needs either EgressSourceIP or EgressGateway set")
+	}
```
Comment (pchaigno) [inline pkg/egressgateway/policy.go]: Do we need this? IIUC, we already have a check in the CRD schema.

File: pkg/egressgateway/policy.go [hunk 20]
```diff
@@ -62,6 +222,29 @@ func ParsePolicy(cenp *v2alpha1.CiliumEgressNATPolicy) (*PolicyConfig, error) {
...
+
+	var policyGwc *policyGatewayConfig
+
+	if egressGateway != nil {
+		if egressGateway.Interface != "" && egressGateway.EgressIP != "" {
+			return nil, fmt.Errorf("CiliumsEgressNATPolicy's gateway configuration can't specify both an interface and an egress IP")
+		}
```
Comment (pchaigno) [inline pkg/egressgateway/policy.go:230]: Couldn't we prevent this as part of the CRD schema?

File: pkg/egressgateway/manager.go (hunk 18 above)
Comment (jibi) [reply pkg/egressgateway/manager.go:414]: (synced offline)

File: bpf/bpf_host.c [hunk 21]
```diff
@@ -1131,6 +1131,79 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
...
+			enum ct_status ct_status;
+
+			/* If the packet is destined to an entity inside the cluster,
+			 * either EP or node, it should not be forwarded to an egress
+			 * gateway since only traffic leaving the cluster is supposed to
+			 * be masqueraded with an egress IP.
+			 */
+			if (is_cluster_destination(ip4, remote_ep->sec_label,
+						   remote_ep->tunnel_endpoint))
+				goto skip_egress_gateway;
```
Comment (pchaigno) [inline bpf/bpf_host.c]: AFAIU, at this point we know the packet isn't going to a local endpoint so we can remove one map lookup (`lookup_ip4_endpoint`) from `is_cluster_destination` (probably needs a big comment for the function though).

File: bpf/bpf_host.c [hunk 22]
```diff
@@ -1131,6 +1131,79 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
...
+		if (!egress_gw_policy)
+			goto skip_egress_gateway;
+
+		/* If the gateway node is the local node, then just let the
+		 * packet go through, as it will be SNATed later on by
+		 * handle_nat_fwd().
+		 */
+		local_ep = __lookup_ip4_endpoint(egress_gw_policy->gateway_ip);
+		if (local_ep && (local_ep->flags & ENDPOINT_F_HOST))
+			goto skip_egress_gateway;
```
Comment (pchaigno) [inline bpf/bpf_host.c]: If the gateway is the local node, then `remote_ep->sec_label == HOST_ID`, no?

File: bpf/bpf_lxc.c [hunk 23]
```diff
@@ -848,46 +848,6 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
 	}
 #endif /* ENABLE_HOST_FIREWALL && !ENABLE_ROUTING */
 
-#ifdef ENABLE_EGRESS_GATEWAY
```
Comment (pchaigno) [inline bpf/bpf_lxc.c]: Do we still have any egress gateway logic in `bpf_lxc.c`? If not, we can update `bpf/Makefile` and `bpf/complexity-tests/*/bpf_lxc.txt` a bit.

**Review by pchaigno — CHANGES_REQUESTED** [review]

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml (hunk 9 above)
Comment (jibi) [reply pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml:51]: I was going to add it as part of the documentation changes, but I can do it now :+1:

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml (hunk 10 above)
Comment (jibi) [reply pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml]: I was considering deprecating it, but I wanted to get some feedback first before actually marking it as such

File: pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml (hunk 11 above)
Comment (jibi) [reply pkg/k8s/apis/cilium.io/client/crds/v2alpha1/ciliumegressnatpolicies.yaml]: Sorry, the comment is wrong here :disappointed: `egressIP` is just the IP used to masquerade traffic, but we'll use the node internal IP to forward traffic to the gateway node (given that we have a `NodeSelector` that identifies it and so we can use that to get its internal IP). I'll fix the description

File: Documentation/gettingstarted/egress-gateway.rst (hunk 13 above)
Comment (jibi) [reply Documentation/gettingstarted/egress-gateway.rst]: This whole PR is not going to be backported to v1.11 (and this change slipped through but should go with the rest of the docs changes). I'll remove it for now

File: pkg/egressgateway/manager.go (hunk 14 above)
Comment (jibi) [reply pkg/egressgateway/manager.go:225]: It's not required as we could do everything with just the `manager.nodes` slice, but it simplifies the code a bit whenever we add or remove a node (as we just add/remove nodes to the map and then sort the list we derive from that map)

File: pkg/egressgateway/policy.go (hunk 19 above)
Comment (jibi) [reply pkg/egressgateway/policy.go]: Yes, I just switched the CRD to use `OneOf` but forgot to remove this

File: pkg/egressgateway/policy.go (hunk 20 above)
Comment (jibi) [reply pkg/egressgateway/policy.go:230]: I don't think we can validate this in the CRD though as `OneOf` semantic is "exactly one", while in this case we allow the user to not specify anything (and just use the first IPv4 assigned to the interface with the default route)

File: bpf/bpf_host.c (hunk 21 above)
Comment (jibi) [reply bpf/bpf_host.c]: The problem is that that function is used also in `snat_v4_needed()`, I need to check if it's possible to refactor it

File: bpf/bpf_host.c (hunk 22 above)
Comment (jibi) [reply bpf/bpf_host.c]: `remote_ep` refers to the EP associated with the actual destination IP of the packet (while here we are looking up the gateway IP which should be a node internal IP)

Comment (pchaigno) [reply bpf/bpf_host.c]: Oh, of course :facepalm: 

I still wonder if there's a more efficient way we could check this. Once we remove `egressIP` in the CRD can `gateway_ip` still be any IP address assigned to the gateway node or will it only be the InternalIP (that used for tunneling)?

File: pkg/egressgateway/net.go [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+
+var zeroIPv4Net = &net.IPNet{IP: net.ParseIP("0.0.0.0"), Mask: net.CIDRMask(0, 32)}
+
+func getIfaceFirstIPv4Address(ifaceName string) (net.IPNet, int, error) {
+	dev, err := netlink.LinkByName(ifaceName)
+	if err != nil {
+		return net.IPNet{}, 0, err
+	}
+
+	addrs, err := netlink.AddrList(dev, netlink.FAMILY_ALL)
```
Comment (julianwiedmann) [inline pkg/egressgateway/net.go]: Would we benefit from

```suggestion
	addrs, err := netlink.AddrList(dev, netlink.FAMILY_V4)
```

File: pkg/egressgateway/net.go [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+func getIfaceWithIPv4Address(ip net.IP) (string, int, net.IPMask, error) {
+	links, err := netlink.LinkList()
+	if err != nil {
+		return "", 0, nil, err
+	}
+
+	for _, l := range links {
+		addrs, err := netlink.AddrList(l, netlink.FAMILY_ALL)
```
Comment (julianwiedmann) [inline pkg/egressgateway/net.go]: Also

```suggestion
		addrs, err := netlink.AddrList(l, netlink.FAMILY_V4)
```

?

File: pkg/egressgateway/policy.go [hunk 26]
```diff
@@ -44,6 +81,129 @@ func (config *PolicyConfig) selectsEndpoint(endpointInfo *endpointMetadata) bool
...
+	case gc.egressIP != nil && !gc.egressIP.Equal(net.IPv4zero):
+		// If the gateway config specifies an egress IP, use the interface with that IP as egress
+		// interface
+		gwc.egressIP.IP = gc.egressIP
+		gwc.ifaceName, gwc.ifaceIndex, gwc.egressIP.Mask, err = getIfaceWithIPv4Address(gc.egressIP)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve interface with egress IP: %w", err)
+		}
+	default:
+		// If the gateway config doesn't specify any egress IP or interface, us
```
Comment (julianwiedmann) [inline pkg/egressgateway/policy.go]:
```suggestion
		// If the gateway config doesn't specify any egress IP or interface, use
```

File: pkg/egressgateway/policy.go [hunk 27]
```diff
@@ -10,15 +10,48 @@ import (
...
+	iface        string
+	egressIP     net.IP
+}
+
+// gatewayConfig is the gateway configuration derived at runtime from a policy.
+//
+// Some of these fields are derived from the running system as the policy may
+// specify only the egress IP (and so we need to figure out which interface has
+// that IP assigned to) or the interface (and in this case we need to find the
+// first IPv4 assigned to that).
```
Comment (julianwiedmann) [inline pkg/egressgateway/policy.go:37]: I think we might want to point this out in the docs as well - any subsequent change to a node's egress interface config won't be reconciled with the policy.

File: bpf/bpf_host.c [hunk 28]
```diff
@@ -1131,6 +1131,79 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
 					      METRIC_EGRESS);
 #endif /* ENABLE_HOST_FIREWALL */
 
+#if defined(ENABLE_EGRESS_GATEWAY) && defined(ENABLE_IPV4)
+	{
+		void *data, *data_end;
```
Comment (brb) [inline bpf/bpf_host.c]: I think this code block is big enough to be moved to a separate function.

Also, JFYI we are adding the WG blob too to the similar place - https://github.com/cilium/cilium/pull/19407/files. The egress blob should go before the WG blob, as the redirect for the encryption will be handled by  `encap_and_redirect_with_nodeid()`. Just that the egress blob should be able to handle `CTX_ACT_REDIRECT`.

File: bpf/bpf_host.c (hunk 21 above)
Comment (jibi) [reply bpf/bpf_host.c]: Let's recap what's going on:
* we are calling `remote_ep = lookup_ip4_remote_endpoint(ip4->daddr)`, which looks up the ep in the ipcache map
* we are passing `remote_ep->sec_label` and `remote_ep->tunnel_endpoint` to `is_cluster_destination()`
* in `is_cluster_destination()` we are calling `lookup_ip4_endpoint(ip4)` (i.e. `__lookup_ip4_endpoint(ip4->daddr)`) which looks up the ep in the endpoint map

To my understanding (and after some tests/printks):
* `lookup_ip4_remote_endpoint()` will return a `struct remote_endpoint_info *` both for local and remote endpoints
* `lookup_ip4_endpoint()` will only return a `struct endpoint_info *` if the endpoint is local

so we could just pass `remote_ep` to `is_cluster_destination()` and get rid of the `lookup_ip4_endpoint(ip4)` check (if `remote_ep` is NULL it means that's not a local/remote ep, so the `lookup_ip4_endpoint(ip4)` case would be covered)

edit: `remote_ep` is never going to be NULL, so we would need to check the actual sec label (see below)

Comment (jibi) [reply bpf/bpf_host.c]: Actually, could we simplify that logic even more and just check `remote_ep->sec_label`? 

```c
static __always_inline bool
is_cluster_destination(struct remote_endpoint_info *remote_ep)
{
    if (!remote_ep)
        return false; // just to keep the verifier happy as the ipcache is always going to return at least the `0.0.0.0/0` entry
    if (remote_ep->sec_labels == WORLD_ID)
        return false

    return true
}
```

File: pkg/datapath/linux/linux_defaults/linux_defaults.go [hunk 29]
```diff
@@ -57,6 +64,12 @@ const (
 	// RulePriorityWireguard is the priority of the rule used for routing packets to Wireguard device for encryption
 	RulePriorityWireguard = 1
 
+	// RulePriorityEgressGateway is the priority used in IP routes added by the manager.
+	// This value was picked as it's lower then the ones used by Cilium
```
Comment (qmonnet) [inline pkg/datapath/linux/linux_defaults/linux_defaults.go]:
```suggestion
	// This value was picked as it's lower than the ones used by Cilium
```

File: bpf/bpf_host.c (hunk 22 above)
Comment (jibi) [reply bpf/bpf_host.c]: Once `egressSourceIP` is removed the manager should only assign internal IPs to the `gateway_ip` field in the policy map

File: pkg/egressgateway/manager.go (hunk 15 above)
Comment (jibi) [reply pkg/egressgateway/manager.go:359]: Done :+1:

File: bpf/bpf_host.c (hunk 22 above)
Comment (pchaigno) [reply bpf/bpf_host.c]: Ok, so when that happens, we should be able to avoid one map lookup here by instead checking if `egress_gw_policy->gateway_ip == INTERNAL_NODE_IP`.

**Review by pchaigno — APPROVED** [review]
Thanks for the new refactoring commit! :rocket:

**Comment by jibi** [comment]
rebased on top of #19194

**Comment by jibi** [comment]
/test-1.23-net-next

**Comment by jibi** [comment]
/test-1.23-net-next

**Review by qmonnet — APPROVED** [review]
Sorry for the late review. Looks all good to me, nice work!

Just two (non-blocking) comments:
- Should the new agent option be documented in Documentation/operations/upgrade.rst (especially now that egress gateway is stable)?
- Maybe a matter of personal preferences, but I found it would be nice to have a bit more context/motivations in some commit logs that you left empty, so that we can understand better the rationale behind the changes when looking at the history.

**Comment by jibi** [comment]
/test-1.23-net-next

**Comment by jibi** [comment]
/test-1.23-net-next

**Comment by jibi** [comment]
/test-1.23-net-next

File: pkg/egressgateway/policy.go [hunk 30]
```diff
@@ -54,13 +214,27 @@ func ParsePolicy(cenp *v2alpha1.CiliumEgressNATPolicy) (*PolicyConfig, error) {
...
 		return nil, fmt.Errorf("CiliumEgressNATPolicy must have a name")
 	}
 
 	egressIP := net.ParseIP(cenp.Spec.EgressSourceIP).To4()
+	egressGateway := cenp.Spec.EgressGateway
+
+	var policyGwc *policyGatewayConfig
+	if egressGateway != nil {
+		if egressGateway.Interface != "" && egressGateway.EgressIP != "" {
+			return nil, fmt.Errorf("CiliumsEgressNATPolicy's gateway configuration can't specify both an interface and an egress IP")
```
Comment (julianwiedmann) [inline pkg/egressgateway/policy.go]: Nitpick:
```suggestion
			return nil, fmt.Errorf("CiliumEgressNATPolicy's gateway configuration can't specify both an interface and an egress IP")
```

**Review by julianwiedmann — APPROVED** [review]

**Comment by pchaigno** [comment]
How did I let this through? There's no explanation for why commit `egressgw: add support for ENI IP rules/routes` is needed :disappointed:

File: pkg/egressgateway/net.go [hunk 31]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	return route.ReplaceRule(ipRule)
+}
+
+func getFirstIPInHostRange(ip net.IPNet) net.IP {
+	out := make(net.IP, net.IPv4len)
+	copy(out, ip.IP.To4())
+	out = out.Mask(ip.Mask)
+	out[3] += 1
```
Comment (pchaigno) [inline pkg/egressgateway/net.go:110]: What's the logic here? If the egress IP is 10.0.1.45/32, then 10.0.1.46 will be selected as the gateway. Why?
